File: TypeScript/app/gilded-rose.ts
export class Item {
  name: string;
  sellIn: number;
  quality: number;

  constructor(name: string, sellIn: number, quality: number) {
    this.name = name;
    this.sellIn = sellIn;
    this.quality = quality;
  }
}

enum ItemType { 
  Normal = 'normal',
  AgedBrie = 'Aged Brie',
  Sulfuras = 'Sulfuras, Hand of Ragnaros',
  Backstage = 'Backstage passes to a TAFKAL80ETC concert',
  Conjured = 'Conjured Mana Cake'
}

class ItemHandler extends Item{
  constructor(name: string, sellIn: number, quality: number) { 
    super(name, sellIn, quality);
  }

  addQuality(count: number) {
    this.quality += count;
  }
  minusQuality(count: number) { 
    this.quality -= count;
  }
  minusSellIn() {
    this.sellIn -= 1;
  }
  isAvailableQuality() { 
    return this.quality > 0 && this.quality < 50;
  }
  isSellIn() {
    return this.sellIn < 1;
  }
  updateQuality() {
    if (this.name === ItemType.Sulfuras) return this

    if (this.isSellIn() && this.isAvailableQuality()) {
      if (this.name === ItemType.AgedBrie) {
        this.addQuality(1);
      } 
      if (this.name === ItemType.Backstage) { 
        this.minusQuality(this.quality);
      }
      if ([ItemType.Normal, ItemType.Conjured].includes(this.name as ItemType)) {
        this.minusQuality(this.name === ItemType.Conjured ? 2 : 1);
      }
    }

    if (this.isAvailableQuality()) {
      if (this.name === ItemType.Normal) { 
        this.minusQuality(1);
      }
      if (this.name === ItemType.AgedBrie) { 
        this.addQuality(1);
      }
      if (this.name === ItemType.Backstage) {
        this.addQuality(1);
        if (this.sellIn < 11) {
          this.addQuality(1);
        }
        if (this.sellIn < 6) {
          this.addQuality(1);
        }
      }
      if (this.name === ItemType.Conjured) { 
        this.minusQuality(2);
      }
    }

    this.minusSellIn();

    return this;
  }
}

class Normal extends ItemHandler {
  constructor(sellIn: number, quality: number) {
    super(ItemType.Normal, sellIn, quality);
  }

  updateQuality() {
    return super.updateQuality()
  }
}

class AgedBrie extends ItemHandler {
  constructor(sellIn: number, quality: number) {
    super(ItemType.AgedBrie, sellIn, quality);
  }

  updateQuality() {
    return super.updateQuality()
  }
}

class Sulfuras extends ItemHandler {
  constructor(sellIn: number, quality: number) {
    super(ItemType.Sulfuras, sellIn, quality);
  }

  updateQuality() {
    return super.updateQuality()
  }
}

class Backstage extends ItemHandler {
  constructor(sellIn: number, quality: number) {
    super(ItemType.Backstage, sellIn, quality);
  }

  updateQuality() {
    return super.updateQuality()
  }
}

class Conjured extends ItemHandler {
  constructor(sellIn: number, quality: number) {
    super(ItemType.Conjured, sellIn, quality);
  }

  updateQuality() {
    return super.updateQuality()
  }
}

export class GildedRose {
  items: Array<Item>;

  constructor(items = [] as Array<Item>) {
    this.items = items;
  }

  updateQuality() {
    for (let i = 0; i < this.items.length; i++) {
      const thisItem = this.items[i];

      if (thisItem.name === ItemType.AgedBrie) {
        const agedBrie = new AgedBrie(thisItem.sellIn, thisItem.quality);
        this.items[i] = agedBrie.updateQuality();
      }

      if (thisItem.name === ItemType.Sulfuras) {
        const sulfuras = new Sulfuras(thisItem.sellIn, thisItem.quality);
        this.items[i] = sulfuras.updateQuality();
      }

      if (thisItem.name === ItemType.Normal) {
        const normal = new Normal(thisItem.sellIn, thisItem.quality);
        this.items[i] = normal.updateQuality();
      }

      if (thisItem.name === ItemType.Backstage) {
        const backstage = new Backstage(thisItem.sellIn, thisItem.quality);
        this.items[i] = backstage.updateQuality();
      }

      if (thisItem.name === ItemType.Conjured) {
        const conjured = new Conjured(thisItem.sellIn, thisItem.quality);
        this.items[i] = conjured.updateQuality();
      }
    }

    return this.items;
  }
}

File: csharpcore/GildedRose/GildedRose.cs
using System.Collections.Generic;
namespace GildedRoseKata
{
    public class GildedRose
    {
        IList<Item> Items;
        public GildedRose(IList<Item> Items)
        {
            this.Items = Items;
        }

        public void UpdateQuality()
        {
            foreach(var item in Items)
            {
                //處理后台通行证與陈年布利奶酪
                if (item.Name == "Aged Brie" || item.Name == "Backstage passes to a TAFKAL80ETC concert")
                {

                    item.Quality = item.Quality < 50 ? item.Quality + 1 : item.Quality;
                    //處理后台通行证
                    if (item.Name == "Backstage passes to a TAFKAL80ETC concert")
                    {
                        //天數5<x<10會再增加一
                        if (item.SellIn <= 10)
                        {
                            item.Quality = item.Quality < 50 ? item.Quality + 1 : 50;
                        }
                        //天數x<=5會再增加一
                        if (item.SellIn <= 5)
                        {
                            item.Quality = item.Quality < 50 ? item.Quality + 1 : 50;
                        }
                        if (item.SellIn <= 0)
                        {
                            item.Quality = 0;
                        }
                    }
                    else
                    {
                        //處理陳年乳酪過了銷售日品質會再加一
                        if(item.SellIn<=0)
                            item.Quality = item.Quality < 50 ? item.Quality + 1 : 50;
                    }
                    
                    item.SellIn--;

                }
                else if (item.Name == "Sulfuras, Hand of Ragnaros")//不處理傳奇物件
                {
                    continue;
                }
                else//其餘商品 
                {
                    //品質大於零時每天減少一點
                    item.Quality = item.Quality > 0 ? item.Quality - 1 : 0;
                    if (item.Name == "Conjured Mana Cake")//新需求加入cake
                    {
                        item.Quality = item.Quality > 0 ? item.Quality - 1 : 0;
                    }
                    item.SellIn--;
                    if (item.SellIn < 0 && item.Quality > 0)
                    {
                        item.Quality = item.Quality - 1;
                        if (item.Name == "Conjured Mana Cake" && item.Quality>0)//新需求加入cake品質都會增加一
                        {
                            item.Quality = item.Quality - 1;
                        }
                    }
                    
                }
                
            }
        }
    }
}
